lib/reversetunnel/agent_test.go

```diff
@@ -18,89 +18,251 @@ package reversetunnel
+		return m.MockSendRequest(name, wantReply, payload)
+	}
+	return false, nil, trace.NotImplemented("")
+}
+
+func (m *mockSSHClient) OpenChannel(name string, data []byte) (ssh.Channel, <-chan *ssh.Request, error) {
+	if m.MockOpenChannel != nil {
+		return m.MockOpenChannel(name, data)
+	}
+	return nil, nil, trace.NotImplemented("")
+}
+
+func (m *mockSSHClient) Close() error {
+	if m.MockClose != nil {
+		return m.MockClose()
+	}
+	return nil
+}
+
+func (m *mockSSHClient) Wait() error {
+	return nil
+}
+
+func (m *mockSSHClient) Principals() []string {
+	return m.MockPrincipals
+}
+
+func (m *mockSSHClient) HandleChannelOpen(channelType string) <-chan ssh.NewChannel {
+	return m.MockHandleChannelOpen
+}
+
+func (m *mockSSHClient) Reply(r *ssh.Request, ok bool, payload []byte) error {
+	if m.MockReply != nil {
+		return m.MockReply(r, ok, payload)
+	}
+
+	return trace.NotImplemented("")
+}
+
+func (m *mockSSHClient) GlobalRequests() <-chan *ssh.Request {
+	return m.MockGlobalRequests
+}
+
+type mockSSHChannel struct {
+	MockSendRequest func(name string, wantReply bool, payload []byte) (bool, error)
+}
+
+func (m *mockSSHChannel) Read(data []byte) (int, error) {
+	return 0, trace.NotImplemented("")
+}
+
+func (m *mockSSHChannel) Write(data []byte) (int, error) {
+	return 0, trace.NotImplemented("")
+}
+
+func (m *mockSSHChannel) Close() error { return nil }
+
+func (m *mockSSHChannel) CloseWrite() error { return nil }
+
+func (m *mockSSHChannel) SendRequest(name string, wantReply bool, payload []byte) (bool, error) {
+	if m.MockSendRequest != nil {
+		return m.MockSendRequest(name, wantReply, payload)
+	}
+
+	return false, trace.NotImplemented("")
+}
+
+func (m *mockSSHChannel) Stderr() io.ReadWriter {
+	return nil
+}
 
-	_, err = agent.connect()
-	require.Error(t, err, "agent should reject invalid host certificate")
+// mockAgentInjection impements several interaces for injecting into an agent.
+type mockAgentInjection struct {
+	client SSHClient
 }
 
-type fakeClient struct {
-	auth.Client
-	caKey ssh.PublicKey
+func (m *mockAgentInjection) transport(context.Context, ssh.Channel, <-chan *ssh.Request, ssh.Conn) *transport {
+	return &transport{}
 }
 
-func (fc *fakeClient) GetCertAuthorities(ctx context.Context, caType types.CertAuthType, loadKeys bool, opts ...services.MarshalOption) ([]types.CertAuthority, error) {
-	ca, err := types.NewCertAuthority(types.CertAuthoritySpecV2{
-		Type:         types.HostCA,
-		ClusterName:  "example.com",
-		CheckingKeys: [][]byte{ssh.MarshalAuthorizedKey(fc.caKey)},
+func (m *mockAgentInjection) DialContext(context.Context, utils.NetAddr) (SSHClient, error) {
+	return m.client, nil
+}
+
+func testAgent(t *testing.T) (*Agent, *mockSSHClient) {
+	tracker, err := track.New(context.Background(), track.Config{
+		ClusterName: "test",
 	})
-	if err != nil {
-		return nil, trace.Wrap(err)
+	require.NoError(t, err)
+
+	addr := utils.NetAddr{Addr: "test-proxy-addr"}
+	tracker.Start()
+
+	lease := <-tracker.Acquire()
```

Comment:
should this be in a switch with a case checking for a timeout so this doesn't unintentionally block tests forever?

Reply:
This would only happen if the tracker implementation changes. I am going to leave it for now but fine with changing this if you feel its necessary.